### ISSUE TITLE
Macro expansion doesn't work correctly when formal argument names clash with variable names using in actual argument expressions

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#208]
Example:

```
#!silver
method test()
    ensures (
        true
        --*
        let tmp2 == (
            true
            ) in (
            tmp2
            --*
            true
        )
    )
{
}
```

Since this example crashes both back-ends with `NoSuchElementException: key not found: tmp2`, I suspect that this is a Silver issue. This is a blocking issue for Prusti.

[Issue created by vakaras: 2018-03-18]
[Last updated on bitbucket: 2018-03-19]

[Comment created by vakaras: 2018-03-19]
Potentially fixed in 1ad051dc67f233343996d7c42e57d63b9ec0ae81. @mschwerhoff: could you please take a look if that commit makes sense to you?

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-06-19 19:01

@alexander_summers That's exactly what happened; I'll fix this as soon as possible.

Somewhat related: we decided to rename bound variables in macro bodies to avoid name clashes, e.g. the forall-bound variable `i` in

    define M(a) forall i: Int :: f(a, i)

will be given a fresh name every time `M` is expanded. If `M(i)` is expanded, the passed-in `i` could now also be renamed, yielding

    forall i0: Int :: f(i0, i0)

or it could not be renamed, yielding

    forall i0: Int :: f(i, i0)

The latter seems more intuitive to me; the corresponding policy would be that actual arguments are not affected by renamings since only variables bound inside the macro are to be renamed.

Do you agree?

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-06-19 21:25

Hi Malte. Yes - the latter is the correct behaviour - the former is basically variable capture, which should be avoided. This can also be done lazily, by instead making the substitution operation itself generate renamings on the fly when necessary. That might be cleaner in that it would avoid renamings when not needed, but could be more work.

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-06-20 14:09

Resolved with commit a488107eef0e
